rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,166 @@
+
+At the time of writing the Policy Server is configured with a ConfigMap that contains all the policies to be loaded.
+When a new policy is created, the ConfigMap is updated, and the Policy Server is restarted to load the new policy.
+
+We want to change this behavior to allow the Policy Server to act as a controller that can load new policies without requiring a restart.
+The Policy Server will be watching the Policy CRD directly and it will change the internal state of the evaluation environment when a new policy is created or updated.
+
+A leader election mechanism will be implemented to ensure that only one Policy Server instance is acting as a controller at a time.
+The leader will be responsible to pull the policies from the registry, validate the policy settings, precompile the policy and store it in a shared cache.
+After the policy is precompiled, the leader will change the status of the policy to Valid and signal the other replicas that a new policy is available to be loaded.
+The replicas will be watching the Policy CRDs and load the new policy when the status is `Valid`.
+In case of an error, the policy status will be updated, but the Policy Server will continue serving the previous valid policy until the issue is resolved.
+
+### Shared Policy Cache
+
+The Policy Server will have a shared cache where the precompiled policies will be stored.
+We can use a RWX Persistent Volume to store the cache and share it among all the Policy Server replicas.
+
+### Policy Versioning
+
+We can use the `generation` field of the Policy CRD to version policies. When a new policy is created or updated, Kubernetes automatically increments the `generation` field.
+
+The current policy will be stored in another CRD, similar to the `Policy` CRD, but with a different name, such as `PolicyReplica`.
+Only one `PolicyReplica` will be kept at any time. When a Policy Server replica restarts or a new replica is created, it will load the policy from the `PolicyReplica` CRD.
+
+The `PolicyReplica` is always assumed to be valid, as it has already been validated by the leader.
+If a Policy Server replica fails to load the policy from the `PolicyReplica` CRD, it will exit with an error.
+Since Kubernetes removes the pod from the service endpoints during a restart or a termination, the Policy Server service will remain consistent, ensuring that no server operates without a valid policy.
+
+Additionally, the policy validation endpoint will include the `generation` number in its URL.
+When all Policy Server replicas are ready with the new `PolicyReplica`, the validation endpoint will switch to the new `generation` number.
+Once this transition is complete, the `PolicyReplica` (current state) and the `Policy` (desired state) will be identical, ensuring consistency across all Policy Server replicas.
+This approach guarantees that the Policy Server will always have a valid policy to load and avoids any service interruptions during updates.
```

Comment:
I'm fine with that, I think it's nice. However, I think we need to have more than 1 `PolicyInstance` object per Policy.

Ideally we would need 2 objects. The `PolicyInstance` would have a boolean field `active` and only the ones with `active == true` are going to be served by the Policy Servers.

Let's go through a timeline of events (which is something we should also make part of the RFC):

- User creates a `AdmissionPolicy`
- The leader of the Policy Servers notices that, downloads, optimize and verify the policy.
- The leader creates a `AdmissionPolicyInstance` object with `gen: 1` and `active: true`
- All the policy servers expose the policy as `/validate/<id of the policy>/g1`
- The webhook is created and points to `/validate/<id of the policy>/g1`

The user makes a change to the previously created policy:
- The leader notices the change, downloads, optimizes and verifies the policy
- The leader creates a new `AdmissionPolicyInstance` with `g2` and `active: true`
- All the policy servers expose the new instance of the policy as `/validate/<id of the policy>/g2`. The `/validate/<id of the policy>/g1` is still active
- The kubewarden controller monitors the rollout of the g2 version of the policy. Once this is available on all the instances it updates `AdmissionPolicyInstance(g1)` to have `active: false`
- The policy servers notice the change stop serving the `g1` policy

The user can eventually do a rollback, but this is something we need to think more about...

Thinking more about that, we might not even need the `active